src/cutadapt/_kmer_finder.pyx

```diff
@@ -0,0 +1,303 @@
+                populate_needle_mask(self.search_masks + mask_offset, search_word, offset - 1,
+                                     self.ref_wildcards, self.query_wildcards)
+        self.positions_and_kmers = positions_and_kmers
+
+    def __reduce__(self):
+        return KmerFinder, (self.positions_and_kmers, self.ref_wildcards, self.query_wildcards)
+
+    def kmers_present(self, str sequence):
+        cdef:
+            KmerSearchEntry entry
+            size_t i
+            size_t kmer_offset
+            bitmask_t zero_mask
+            bitmask_t found_mask
+            ssize_t start, stop
+            const bitmask_t *mask_ptr
+            const char *search_ptr
+            bint search_result
+            ssize_t search_length
+        if not PyUnicode_IS_COMPACT_ASCII(sequence):
+            raise ValueError("Only ASCII strings are supported")
+        cdef const char *seq = <char *>PyUnicode_DATA(sequence)
+        cdef Py_ssize_t seq_length = PyUnicode_GET_LENGTH(sequence)
+        for i in range(self.number_of_searches):
+            entry = self.search_entries[i]
+            start = entry.search_start 
+            stop = entry.search_stop
+            if start < 0:
+                start = seq_length + start
+                if start < 0:
+                    start = 0
+            elif start > seq_length:
+                continue
+            if stop < 0:
+                step = seq_length + stop
+                if stop <= 0:  # No need to search
+                    continue
+            elif stop == 0:
+                stop = seq_length
+            search_length = stop - start
+            if search_length <= 0:
+                continue
+            search_ptr = seq + start
+            zero_mask = entry.zero_mask
+            found_mask = entry.found_mask
+            mask_ptr = self.search_masks + entry.mask_offset
+            search_result = shift_or_multiple_is_present(
+                search_ptr, search_length, mask_ptr, zero_mask, found_mask)
+            if search_result:
+                return True
+        return False
+
+    def __dealloc__(self):
+        PyMem_Free(self.search_masks)
+        PyMem_Free(self.search_entries)
+
+
+cdef void set_masks(bitmask_t *needle_mask, size_t pos, const char *chars):
+    cdef char c
+    cdef size_t i
+    for i in range(strlen(chars)):
+        needle_mask[<uint8_t>chars[i]] &= ~(<bitmask_t>1ULL << pos)
+
+cdef populate_needle_mask(bitmask_t *needle_mask, const char *needle, size_t needle_length,
+                          bint ref_wildcards, bint query_wildcards):
+    cdef size_t i
+    cdef char c
+    cdef uint8_t j
+    if needle_length > (sizeof(bitmask_t) * 8 - 1):
+        raise ValueError("The pattern is too long!")
+    memset(needle_mask, 0xff, sizeof(bitmask_t) * BITMASK_INDEX_SIZE)
+    for i in range(needle_length):
+        c = needle[i]
+        if c == 0:
+            continue
+        if c == b"A" or c == b"a":
```

Comment:
There must be a simpler way to do this. The string constants hide what is going on (I only saw on second glance that they are all different, which of course they must be, but still.)

I’d prefer if the mapping of IUPAC wildcards to which characters they represent were encoded in one place and if that representation were used to set the masks as in [`_align.pyx`](https://github.com/marcelm/cutadapt/blob/main/src/cutadapt/_align.pyx#L47). It doesn’t even have to be in a `cdef` function; this is only run once after all.

Reply:
Yes, let me think about this one. I should do this before your next review.

Reply:
Much nicer, thanks!